File: ROUN1B/seg.py
T = int(input())
if T<1 or T>1000:
    exit(0)
sum = 0
for t in range (T):
    M1 = []
    M2 = []
    L = []
    R = []
    P = []
    N,K = [int(x) for x in input().split()]
    sum += N
    if K<1 or N<1 or K>10**5 or N>10**5 or sum>(5 * 10**5) or K>N:
        exit(0)
    for i in range(N):
        a,b = [int(x) for x in input().split()]
        M1.append(a)
        M2.append(b)
        if 1>a or 1>b or a>b or a>10**9 or b>10**9:
            exit(0)
        L.append(a)
        R.append(b)
    L.sort()
    R.sort()
    j = 0
    k = 0
    for i in range (2*N):
        if k<N and j<N:
            if R[k]>L[j]:
                t = [L[j],"S"]
                P.append(t)
                j+=1
            else:
                ind = M2.index(R[k])
                t = [R[k],"E",M1[ind]]
                del(M1[ind])
                del(M2[ind])
                P.append(t)
                k+=1
        elif j<N and k>=N:
            t = [L[j],"S"]
            P.append(t)
            j+=1
        else:
            ind = M2.index(R[k])
            t = [R[k],"E",M1[ind]]
            del(M1[ind])
            del(M2[ind])
            P.append(t)
            k+=1
    print(P)
    count = 0
    dis = 0
    curr = []
    dict = {}
    for i in range(2*N):
        if P[i][1] == 'S':
            count+=1
            curr.append(P[i][0])
            dict.setdefault(P[i][0],[]) #creates a key iff the key doesn't already exist
            dict[P[i][0]].append(count)
        else:
            count -= 1
            if count>= K-1:
                dis = max(dis, P[i][0]-curr[-1])
                print(curr)
                print(dis)
                x = curr.index(P[i][2])
                for j in dict[curr[x]]:
                    if j>=K:
                        if P[i][0]-j>dis:
                            dis = P[i][0] - j
                            del[dict[curr[x]][dict[curr[x]].index(j)]]
                del(curr[x])
    print(dict)

File: ROUN1B/seg.cpp
#include<iostream>
#include<stdlib.h>
#include<vector>
#include<algorithm>
main(){
    std::ios::sync_with_stdio(0);
    std::cin.tie(NULL);
    long sum = 0;
    int T, t;
    long N,K,i,M,count,j,k,S;
    std::cin>>T;
    if (T<1||T>1000)
        exit(0);
    for (t=0;t<T;t++){
        std::cin>>N>>K;
        sum += N;
        if (K<1||N<1||K>100000||N>100000||sum>500000||K>N)
            exit(0);
        std::vector<long> L;
        std::vector<long> R;
        for(i = 0; i<N;i++){
            std::cin>>M>>count;
            if (1>M||1>count||M>count||M>1000000000||count>1000000000)
                exit(0);
            L.push_back(M);
            R.push_back(count);
        }
    std::sort(L.begin(),L.end());
    std::sort(R.begin(),R.end());
    j = 0; k = 0; count =0; 
    std::vector<long>start;
    std::vector<long>end;
    for (i = L[0];i<=R[N-1];i++){
        while(L[j]==i && j<N){
            count++;j++;
            if (count >= K){
                start.push_back(i);
                //std::cout<<"yy"<<i<<"\t";
            }
        }
        //std::cout<<count<<"\n";
        while(R[k]==i && k<N){
            count--;k++;
            if(count>= K-1){
                end.push_back(i);
                //std::cout<<"xx"<<i<<"\t";
            }
            S = start.size();
        }
    }
    M = 0;
    //std::reverse(end.begin(),end.end());
    //for (long a:start)
    //    std::cout<<a<<"\t";
    //std::cout<<"\n";
    //for(long a:end)
    //    std::cout<<a<<"\t";
    if (K%2==0)
        for(j=0;j<S;j++)
            M = std::max(M,end[j]-start[j]);
    else
        for(j=0;j<S;j++)
            M = std::max(M,end[S-j-1]-start[j]);
    std::cout<<M<<"\n";
    }
}